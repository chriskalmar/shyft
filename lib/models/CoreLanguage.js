
import {
  dataTypes,
  Entity,
} from '../../';


const {
  DataTypeString,
} = dataTypes



export const CoreLanguage = new Entity({
  name: 'language',
  domain: 'core',
  description: 'A language',

  // indexing: {
  //   unique: [
  //     [ 'name' ],
  //     [ 'iso_code' ],

  attributes: {

    name: {
      type: DataTypeString,
      description: 'The name of the language',
      maxLength: 30,
      required: true
    },

    nativeName: {
      type: DataTypeString,
      description: 'The native name of the language',
      maxLength: 30,
      required: true
    },

    isoCode: {
      type: DataTypeString,
      description: 'ISO code of the language',
      pattern: '^[a-z]+$',
      minLength: 2,
      maxLength: 2,
      required: true
    },
  }

})

