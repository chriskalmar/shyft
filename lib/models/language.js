
'use script';

module.exports = {

  domain: 'core',

  entities: [

    {
      name: 'language',
      description: 'a language',

      cascadeOnDelete: true,

      attributes: [

        {
          name: 'name',
          type: 'string',
          description: 'Name',
          maxLength: 30,
          required: true
        },

        {
          name: 'native_name',
          type: 'string',
          description: 'Name',
          maxLength: 30,
          required: true
        },

        {
          name: 'iso_code',
          type: 'string',
          description: 'ISO 639-1 code of the language',
          maxLength: 2,
          required: true
        },

      ],

      indexing: {
        unique: [
          [ 'name' ],
          [ 'iso_code' ],
        ]
      }

    }

  ]

}
