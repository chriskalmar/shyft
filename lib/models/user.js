
'use script';

module.exports = {

  domain: 'core',

  entities: [

    {
      name: 'user',
      description: 'a user',

      isUserTable: true,
      cascadeOnDelete: true,

      attributes: [

        {
          name: 'username',
          type: 'string',
          description: 'Username',
          pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
          minLength: 3,
          maxLength: 30,
          required: true
        },

        {
          name: 'first_name',
          type: 'string',
          description: 'First name',
          maxLength: 30
        },

        {
          name: 'last_name',
          type: 'string',
          description: 'Last name',
          maxLength: 30
        },

        {
          name: 'language',
          type: 'reference',
          description: 'User language',
          target: 'core::language',
          required: true
        },

        {
          name: 'email',
          type: 'email',
          description: 'Email address',
          required: true
        },

        {
          name: 'password',
          type: 'password',
          description: 'User password',
          required: true
        }
      ],

      indexing: {
        unique: [
          [ 'username' ],
          [ 'email' ],
        ]
      }

    }

  ]

}
