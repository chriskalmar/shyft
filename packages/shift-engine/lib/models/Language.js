
import Entity from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import Index, { INDEX_UNIQUE } from '../index/Index';


export const Language = new Entity({
  name: 'Language',
  description: 'A language',

  indexes: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'name' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'isoCode' ],
    }),
  ],

  attributes: {

    name: {
      type: DataTypeString,
      description: 'The name of the language',
      required: true
    },

    nativeName: {
      type: DataTypeString,
      description: 'The native name of the language',
      required: true
    },

    isoCode: {
      type: DataTypeString,
      description: 'ISO code of the language',
      required: true
    },
  }

})

