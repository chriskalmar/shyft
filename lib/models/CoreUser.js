
import {
  datatypes,
  Entity,
} from '../../';

import CoreLanguage from './CoreLanguage';

const {
  DataTypeString,
} = datatypes

export const CoreUser = new Entity({
  name: 'user',
  domain: 'core',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  // indexing: {
  //   unique: [
  //     [ 'username' ],
  //     [ 'email' ],


  attributes: {

    username: {
      type: DataTypeString,
      description: 'Username',
      pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
      minLength: 3,
      maxLength: 30,
      required: true
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
      maxLength: 30
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
      maxLength: 30
    },

    language: {
      type: CoreLanguage,
      description: 'User language',
      required: true
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true
    }
  }

})



