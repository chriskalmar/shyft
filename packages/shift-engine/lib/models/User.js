
import Entity from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import Index, { INDEX_UNIQUE } from '../index/Index';

import { Language } from './Language';


export const User = new Entity({
  name: 'User',
  domain: 'core',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  indexing: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'username' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'email' ],
    }),
  ],


  attributes: {

    username: {
      type: DataTypeString,
      description: 'Username',
      pattern: '^(?![_.])(?!.*[_.]{2})[a-zA-Z0-9._]+$',
      minLength: 3,
      maxLength: 30,
      required: true
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
      maxLength: 30
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
      maxLength: 30
    },

    language: {
      type: Language,
      description: 'User language',
      required: true
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true
    }
  }

})



