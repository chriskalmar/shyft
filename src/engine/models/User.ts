import { Entity } from '../entity/Entity';
import { DataTypeString } from '../datatype/dataTypes';
import { Index, INDEX_UNIQUE } from '../index/Index';

import { Language } from './Language';

export const User = new Entity({
  name: 'User',
  description: 'A user',

  // isUserTable: true,
  // sequenceGenerator: 'timebased',

  indexes: [
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'username' ],
    }),
    new Index({
      type: INDEX_UNIQUE,
      attributes: [ 'email' ],
    }),
  ],

  attributes: {
    username: {
      type: DataTypeString,
      description: 'Username',
      required: true,
    },

    firstName: {
      type: DataTypeString,
      description: 'First name',
    },

    lastName: {
      type: DataTypeString,
      description: 'Last name',
    },

    language: {
      type: Language,
      description: 'User language',
      required: true,
    },

    email: {
      type: DataTypeString,
      description: 'Email address',
      required: true,
    },

    password: {
      type: DataTypeString,
      description: 'User password',
      required: true,
    },
  },
});
