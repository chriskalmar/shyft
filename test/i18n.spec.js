import './setupAndTearDown';
import {
  mutate,
} from './db';

import {
  asAdmin,
} from './testUtils';

import { Book } from './models/Book';


describe('i18n', () => {

  it('should store translations if provided', async () => {

    const payload = {
      title: 'War and Peace',
      'title.i18n': {
        de: 'Krieg und Frieden'
      },
      author: 'Leo Tolstoy',
    }

    const result = await mutate(Book, 'create', payload, null, asAdmin())
    expect(result).toMatchSnapshot()
  })


  it('should reject translations of unknown languages', async () => {

    const payload = {
      title: 'War and Peace',
      'title.i18n': {
        de: 'Krieg und Frieden',
        fr: 'Guerre et Paix',
      },
      author: 'Leo Tolstoy',
    }

    await mutate(Book, 'create', payload, null, asAdmin())
      .catch(e => {
        expect(e).toMatchSnapshot()
      })
  })


  it('should take main value form translations object if provided', async () => {

    const payload = {
      'title.i18n': {
        default: 'War and Peace',
        de: 'Krieg und Frieden',
      },
      author: 'Leo Tolstoy',
    }

    const result = await mutate(Book, 'create', payload, null, asAdmin())
    expect(result).toMatchSnapshot()
  })


  it('should store translations of multiple attributes', async () => {

    const payload = {
      'title.i18n': {
        default: 'War and Peace',
        de: 'Krieg und Frieden',
      },
      shortSummary: 'The novel chronicles the history of ...',
      'shortSummary.i18n': {
        de: 'Krieg und Frieden ist ein historischer Roman des ...',
      },
      author: 'Leo Tolstoy',
    }

    const result = await mutate(Book, 'create', payload, null, asAdmin())
    expect(result).toMatchSnapshot()
  })


})
