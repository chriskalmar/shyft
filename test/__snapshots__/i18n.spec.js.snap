// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`i18n should ignore translations for unknown languages 1`] = `
Object {
  "author": "Leo Tolstoy",
  "i18n": null,
  "id": "2",
  "shortSummary": null,
  "shortSummary.i18n": Object {
    "de": undefined,
    "en": null,
  },
  "title": "War and Peace",
  "title.i18n": Object {
    "de": "Krieg und Frieden",
    "en": "War and Peace",
  },
}
`;

exports[`i18n should store translations if provided 1`] = `
Object {
  "author": "Leo Tolstoy",
  "i18n": null,
  "id": "1",
  "shortSummary": null,
  "shortSummary.i18n": Object {
    "de": undefined,
    "en": null,
  },
  "title": "War and Peace",
  "title.i18n": Object {
    "de": "Krieg und Frieden",
    "en": "War and Peace",
  },
}
`;

exports[`i18n should store translations of multiple attributes 1`] = `
Object {
  "author": "Leo Tolstoy",
  "i18n": null,
  "id": "4",
  "shortSummary": "The novel chronicles the history of ...",
  "shortSummary.i18n": Object {
    "de": "Krieg und Frieden ist ein historischer Roman des ...",
    "en": "The novel chronicles the history of ...",
  },
  "title": "War and Peace",
  "title.i18n": Object {
    "de": "Krieg und Frieden",
    "en": "War and Peace",
  },
}
`;

exports[`i18n should take main value form translations object if provided 1`] = `
Object {
  "author": "Leo Tolstoy",
  "i18n": null,
  "id": "3",
  "shortSummary": null,
  "shortSummary.i18n": Object {
    "de": undefined,
    "en": null,
  },
  "title": "War and Peace",
  "title.i18n": Object {
    "de": "Krieg und Frieden",
    "en": "War and Peace",
  },
}
`;
